File: superstitioApi/java/superstitioapi/powers/barIndepend/RenderOnThing.java
package superstitioapi.powers.barIndepend;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.megacrit.cardcrawl.core.Settings;
import com.megacrit.cardcrawl.helpers.*;
import com.megacrit.cardcrawl.helpers.input.InputHelper;

import java.util.*;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public abstract class RenderOnThing {
    public static final float HIDE_SPEED = 4.0F;
    public static final float HEALTH_HIDE_TIMER_MIN = 0.2F;
    public static final float BAR_OFFSET_Y = 28.0f * Settings.scale;
    protected static final float TEXT_OFFSET_Y = 11.0f * Settings.scale;
    //绘制相关
    protected static final float BAR_DIAMETER = 20.0f * Settings.scale;
    //    public static final float HITBOX_HEIGHT = ;
//    protected static final float BG_OFFSET_Y = -31.0f * Settings.scale;
    protected final Map<String, AmountChunk> amountChunkWithUuid = new HashMap<>();
    protected final Supplier<Hitbox> hitboxBondTo;
    protected final float HeightOffset;
    public Color barTextColor;
    public Hitbox hitbox;
    public String uuid_self;
    protected float fontScale;
    protected float healthHideTimer = 1.0f;
    protected List<AmountChunk> sortedChunkList = new ArrayList<>();
    private String rawBarText = "%d/%d";

    public RenderOnThing(final Supplier<Hitbox> hitbox, HasBarRenderOnCreature power) {
        this.uuid_self = power.uuidPointTo();
        this.hitboxBondTo = hitbox;
        this.hitbox = new Hitbox(hitboxBondTo.get().width + BAR_DIAMETER * 3f, BAR_DIAMETER * 1.5f);
        updateHitBoxPlace(this.hitbox);

        this.HeightOffset = power.Height();
        this.barTextColor = new Color(1.0f, 1.0f, 1.0f, 1.0f);
    }

    public void removeChunk(HasBarRenderOnCreature hasBarRenderOnCreature) {
        amountChunkWithUuid.remove(hasBarRenderOnCreature.uuidOfSelf());
        reMakeSortedChunkList();
    }

    public void updateHitBoxPlace(Hitbox hitbox) {
        hitbox.move(hitboxBondTo.get().cX,
                HeightOffset + hitboxBondTo.get().cY + hitboxBondTo.get().height / 2);
    }

    public boolean isUuidInThis(String uuid_chunk) {
        return amountChunkWithUuid.containsKey(uuid_chunk);
    }

    protected int getMaxBarAmount() {
        return amountChunkWithUuid.values().stream().mapToInt(chunk -> chunk.maxAmount).max().orElse(10);
    }

    public void render(SpriteBatch sb) {
        renderBarTextWithColorAlphaChange(sb, getXDrawStart(), getYDrawStart());
        if (Settings.isDebug || Settings.isInfo) {
            renderDebug(sb);
        }
    }

    protected void renderDebug(SpriteBatch sb) {
        this.hitbox.render(sb);
    }

    public void update() {
        updateHitBoxPlace(this.hitbox);
        this.hitbox.update();
        for (AmountChunk amountChunk : this.amountChunkWithUuid.values()) {
            amountChunk.update();
        }
        update_showTips(this.hitbox);
        updateHbHoverFade();
    }

    protected final ArrayList<PowerTip> AllTips() {
        return this.amountChunkWithUuid.values().stream().filter(amountChunk -> amountChunk.tip != null)
                .map(amountChunk -> amountChunk.tip).collect(Collectors.toCollection(ArrayList::new));
    }

    protected void update_showTips(Hitbox hitbox) {
        if (hitbox.hovered) {
            renderTip(AllTips());
        }
        this.fontScale = MathHelper.scaleLerpSnap(this.fontScale, 0.7F);
    }

    public void renderTip(ArrayList<PowerTip> tips) {
        if (!((float) InputHelper.mX < 1400.0F * Settings.scale)) {
            TipHelper.queuePowerTips((float) InputHelper.mX - 350.0F * Settings.scale,
                    (float) InputHelper.mY - 50.0F * Settings.scale, tips);
            return;
        }
        TipHelper.queuePowerTips((float) InputHelper.mX + 50.0F * Settings.scale,
                (float) InputHelper.mY + 50.0F * Settings.scale, tips);
    }

    protected abstract float getYDrawStart();

    protected abstract float getXDrawStart();

    /**
     * @param x getXDrawStart
     * @param y getYDrawStart
     */
    protected void renderBarTextWithColorAlphaChange(final SpriteBatch sb, final float x, final float y) {
        final float tmp = this.barTextColor.a;
        this.barTextColor.a *= this.healthHideTimer;
        renderBarText(sb, x, y);
        this.barTextColor.a = tmp;
    }

    /**
     * @param x getXDrawStart
     * @param y getYDrawStart
     */
    protected void renderBarText(SpriteBatch sb, final float x, final float y) {
        FontHelper.renderFontCentered(sb, FontHelper.healthInfoFont, makeBarText(), this.hitbox.cX, y + TEXT_OFFSET_Y,
                this.barTextColor);
    }

    protected final String makeBarText() {
        boolean hasTwoDs = rawBarText.contains("%d") && rawBarText.indexOf("%d") != rawBarText.lastIndexOf("%d");
        boolean hasD = rawBarText.contains("%d");
        if (hasTwoDs)
            return String.format(rawBarText, this.getTotalAmount(), this.getMaxBarAmount());
        if (hasD)
            return String.format(rawBarText, this.getTotalAmount());
        return rawBarText;
    }

    protected void updateHbHoverFade() {
        if (this.hitbox.hovered) {
            this.healthHideTimer -= Gdx.graphics.getDeltaTime() * HIDE_SPEED;
            if (this.healthHideTimer < HEALTH_HIDE_TIMER_MIN) {
                this.healthHideTimer = HEALTH_HIDE_TIMER_MIN;
            }
        } else {
            this.healthHideTimer += Gdx.graphics.getDeltaTime() * HIDE_SPEED;
            if (this.healthHideTimer > 1.0F) {
                this.healthHideTimer = 1.0F;
            }
        }
    }

    /**
     * 在初始化后，如果想要修改，请使用这个函数
     *
     * @param message 消息
     */
    public void tryApplyMessage(BarRenderUpdateMessage message) {
        if (!Objects.equals(this.uuid_self, message.uuidOfBar)) return;
        if (!amountChunkWithUuid.containsKey(message.uuidOfPower)) {
            addNewAmountChunk(message);
            return;
        }

        AmountChunk messageTargetChunk = amountChunkWithUuid.get(message.uuidOfPower);
        this.rawBarText = message.rawTextOnBar;
        messageTargetChunk.applyNoPositionMessage(message);
        if (message.detail != null)
            message.detail.accept(this);
    }

    private void addNewAmountChunk(BarRenderUpdateMessage message) {
        this.amountChunkWithUuid.put(message.uuidOfPower,
                makeNewAmountChunk(message).applyNoPositionMessage(message));
    }

    protected final int getTotalAmount_InFrontOf(int sumToIndex_InFrontOf) {
        if (sumToIndex_InFrontOf <= 0)
            return 0;
        return getChunkOrdered(sumToIndex_InFrontOf - 1).stream().mapToInt(chunk -> chunk.nowAmount).sum();
    }

    protected final List<AmountChunk> getChunkOrdered(int index) {
        reMakeSortedChunkList();
        if (sortedChunkList.size() <= 1)
            return sortedChunkList;
        if (index <= 0)
            return Collections.singletonList(sortedChunkList.get(0));
        return sortedChunkList.subList(0, index);
    }

    protected void reMakeSortedChunkList() {
        List<AmountChunk> orderedList = new ArrayList<>();
        amountChunkWithUuid.values().stream().sorted(AmountChunk::compareTo).forEachOrdered(orderedList::add);
        this.sortedChunkList = orderedList;
    }

    protected final int getTotalAmount() {
        return amountChunkWithUuid.values().stream().mapToInt(value -> value.nowAmount).sum();
    }

    protected abstract AmountChunk makeNewAmountChunk(BarRenderUpdateMessage message);

    protected int getNextOrder() {
        return amountChunkWithUuid.values().stream().mapToInt(value -> value.order).max().orElse(-1) + 1;
    }

    protected abstract static class AmountChunk implements Comparable<AmountChunk> {
        public final int order;
        public PowerTip tip;
        public int maxAmount = 0;
        public int nowAmount = 0;

        public AmountChunk(int order) {
            this.order = order;
        }


        public abstract void render(final SpriteBatch sb);

        public abstract void update();

        public AmountChunk setTip(BarRenderUpdateMessage.ToolTip tip) {
            this.tip = new PowerTip(tip.name, tip.description);
            return this;
        }

        public AmountChunk setMaxAmount(int maxAmount) {
            this.maxAmount = maxAmount;
            return this;
        }

        public AmountChunk setNowAmount(int nowAmount) {
            this.nowAmount = nowAmount;
            return this;
        }

        public abstract AmountChunk applyNoPositionMessage(BarRenderUpdateMessage message);
//            if (message.newAmount != 0 && message.newAmount != this.nowAmount) {
//                this.nowAmount = message.newAmount;
//            }
//            if (message.maxAmount != 0)
//                this.maxAmount = message.maxAmount;
//            if (message.toolTip != null)
//                this.tip = new PowerTip(message.toolTip.name, message.toolTip.description);
//            return this;


        @Override
        public int compareTo(AmountChunk other) {
            return Integer.compare(this.order, other.order);
        }

        public int getOrder() {
            return this.order;
        }
    }

}

File: superstitioApi/java/superstitioapi/powers/barIndepend/RenderOnThing.kt
package superstitioapi.powers.barIndepend

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.megacrit.cardcrawl.core.Settings
import com.megacrit.cardcrawl.helpers.FontHelper
import com.megacrit.cardcrawl.helpers.Hitbox
import com.megacrit.cardcrawl.helpers.MathHelper
import com.megacrit.cardcrawl.helpers.PowerTip
import superstitioapi.powers.barIndepend.BarRenderUpdateMessage.ToolTip
import superstitioapi.utils.TipsUtility
import java.util.function.Supplier

abstract class RenderOnThing(protected val hitboxBondTo: Supplier<Hitbox>, power: HasBarRenderOnCreature)
{
    //    public static final float HITBOX_HEIGHT = ;
    //    protected static final float BG_OFFSET_Y = -31.0f * Settings.scale;
    protected val amountChunkWithUuid: MutableMap<String, AmountChunk> = HashMap()
    protected val HeightOffset: Float
    var barTextColor: Color
    var hitbox: Hitbox = Hitbox(hitboxBondTo.get().width + BAR_DIAMETER * 3f, BAR_DIAMETER * 1.5f)
    var uuid_self: String = power.uuidPointTo()
    protected var fontScale: Float = 0f
    protected var healthHideTimer: Float = 1.0f
    protected var sortedChunkList: List<AmountChunk> = java.util.ArrayList()
    private var rawBarText: String? = "%d/%d"

    init
    {
        updateHitBoxPlace(this.hitbox)

        this.HeightOffset = power.Height()
        this.barTextColor = Color(1.0f, 1.0f, 1.0f, 1.0f)
    }

    open fun removeChunk(hasBarRenderOnCreature: HasBarRenderOnCreature)
    {
        amountChunkWithUuid.remove(hasBarRenderOnCreature.uuidOfSelf())
        reMakeSortedChunkList()
    }

    open fun updateHitBoxPlace(hitbox: Hitbox)
    {
        hitbox.move(
            hitboxBondTo.get().cX,
            HeightOffset + hitboxBondTo.get().cY + hitboxBondTo.get().height / 2
        )
    }

    fun isUuidInThis(uuid_chunk: String?): Boolean
    {
        return amountChunkWithUuid.containsKey(uuid_chunk)
    }

    open fun render(sb: SpriteBatch)
    {
        renderBarTextWithColorAlphaChange(sb, xDrawStart, yDrawStart)
        if (Settings.isDebug || Settings.isInfo)
        {
            renderDebug(sb)
        }
    }

    open fun update()
    {
        updateHitBoxPlace(this.hitbox)
        hitbox.update()
        for (amountChunk in amountChunkWithUuid.values)
        {
            amountChunk.update()
        }
        update_showTips(this.hitbox)
        updateHbHoverFade()
    }

    fun renderTip(tips: MutableList<PowerTip>)
    {
        TipsUtility.renderTipsWithMouse(tips)
    }

    /**
     * 在初始化后，如果想要修改，请使用这个函数
     *
     * @param message 消息
     */
    open fun tryApplyMessage(message: BarRenderUpdateMessage)
    {
        if (uuid_self != message.uuidOfBar) return
        if (!amountChunkWithUuid.containsKey(message.uuidOfPower))
        {
            addNewAmountChunk(message)
            return
        }

        val messageTargetChunk = amountChunkWithUuid[message.uuidOfPower]
        this.rawBarText = message.rawTextOnBar
        messageTargetChunk!!.applyNoPositionMessage(message)
        if (message.detail != null) message.detail!!.accept(this)
    }

    protected val maxBarAmount: Int
        get() = amountChunkWithUuid.values.maxOfOrNull(AmountChunk::maxAmount) ?: 10

    protected open fun renderDebug(sb: SpriteBatch?)
    {
        hitbox.render(sb)
    }

    protected open fun update_showTips(hitbox: Hitbox?)
    {
        if (hitbox!!.hovered)
        {
            renderTip(AllTips())
        }
        this.fontScale = MathHelper.scaleLerpSnap(this.fontScale, 0.7f)
    }

    protected abstract val yDrawStart: Float

    protected abstract val xDrawStart: Float

    /**
     * @param x getXDrawStart
     * @param y getYDrawStart
     */
    protected open fun renderBarTextWithColorAlphaChange(sb: SpriteBatch?, x: Float, y: Float)
    {
        val tmp = barTextColor.a
        barTextColor.a *= this.healthHideTimer
        renderBarText(sb, x, y)
        barTextColor.a = tmp
    }

    /**
     * @param x getXDrawStart
     * @param y getYDrawStart
     */
    protected open fun renderBarText(sb: SpriteBatch?, x: Float, y: Float)
    {
        FontHelper.renderFontCentered(
            sb, FontHelper.healthInfoFont, makeBarText(), hitbox.cX, y + TEXT_OFFSET_Y,
            this.barTextColor
        )
    }

    protected open fun updateHbHoverFade()
    {
        if (hitbox.hovered)
        {
            this.healthHideTimer -= Gdx.graphics.deltaTime * HIDE_SPEED
            if (this.healthHideTimer < HEALTH_HIDE_TIMER_MIN)
            {
                this.healthHideTimer = HEALTH_HIDE_TIMER_MIN
            }
        }
        else
        {
            this.healthHideTimer += Gdx.graphics.deltaTime * HIDE_SPEED
            if (this.healthHideTimer > 1.0f)
            {
                this.healthHideTimer = 1.0f
            }
        }
    }

    protected fun reMakeSortedChunkList()
    {
        val orderedList: MutableList<AmountChunk> = java.util.ArrayList()
        amountChunkWithUuid.values.stream().sorted(AmountChunk::compareTo)
            .forEachOrdered(orderedList::add)
        this.sortedChunkList = orderedList
    }

    protected abstract fun makeNewAmountChunk(message: BarRenderUpdateMessage): AmountChunk?

    protected val nextOrder: Int
        get() = (amountChunkWithUuid.values.maxOfOrNull(AmountChunk::order) ?: -1) + 1

    private fun addNewAmountChunk(message: BarRenderUpdateMessage)
    {
        amountChunkWithUuid[message.uuidOfPower] = makeNewAmountChunk(message)!!.applyNoPositionMessage(message)
    }

    protected fun AllTips(): MutableList<PowerTip>
    {
        return amountChunkWithUuid.values.filter { amountChunk: AmountChunk -> amountChunk.tip != null }
            .mapNotNull(AmountChunk::tip).toMutableList()
    }

    protected fun makeBarText(): String?
    {
        val hasTwoDs = rawBarText!!.contains("%d") && rawBarText!!.indexOf("%d") != rawBarText!!.lastIndexOf("%d")
        val hasD = rawBarText!!.contains("%d")
        if (hasTwoDs) return String.format(
            rawBarText!!,
            totalAmount,
            maxBarAmount
        )
        if (hasD) return String.format(rawBarText!!, this.totalAmount)
        return rawBarText
    }

    protected fun getTotalAmount_InFrontOf(sumToIndex_InFrontOf: Int): Int
    {
        if (sumToIndex_InFrontOf <= 0) return 0
        return getChunkOrdered(sumToIndex_InFrontOf - 1).stream().mapToInt(AmountChunk::nowAmount)
            .sum()
    }

    protected fun getChunkOrdered(index: Int): List<AmountChunk>
    {
        reMakeSortedChunkList()
        if (sortedChunkList.size <= 1) return sortedChunkList
        if (index <= 0) return listOf(sortedChunkList[0])
        return sortedChunkList.subList(0, index + 1)
    }

    protected val totalAmount: Int
        get() = amountChunkWithUuid.values.stream().mapToInt(AmountChunk::nowAmount).sum()

    protected abstract class AmountChunk(val order: Int) : Comparable<AmountChunk>
    {
        var tip: PowerTip? = null
        var maxAmount: Int = 0
        var nowAmount: Int = 0


        abstract fun render(sb: SpriteBatch)

        abstract fun update()

        open fun setTip(tip: ToolTip): AmountChunk
        {
            this.tip = PowerTip(tip.name, tip.description)
            return this
        }

        open fun setMaxAmount(maxAmount: Int): AmountChunk
        {
            this.maxAmount = maxAmount
            return this
        }

        open fun setNowAmount(nowAmount: Int): AmountChunk
        {
            this.nowAmount = nowAmount
            return this
        }

        abstract fun applyNoPositionMessage(message: BarRenderUpdateMessage): AmountChunk

        override fun compareTo(other: AmountChunk): Int
        {
            return this.order.compareTo(other.order)
        }
    }

    companion object
    {
        const val HIDE_SPEED: Float = 4.0f
        const val HEALTH_HIDE_TIMER_MIN: Float = 0.2f
        val BAR_OFFSET_Y: Float = 28.0f * Settings.scale
        val TEXT_OFFSET_Y: Float = 11.0f * Settings.scale

        //绘制相关
        val BAR_DIAMETER: Float = 20.0f * Settings.scale
    }
}